enterprise/frontend/src/embedding-sdk/components/private/Breadcrumb/utils/update-breadcrumbs.ts

```diff
@@ -0,0 +1,46 @@
+import type { SdkBreadcrumbItem } from "embedding-sdk/types/breadcrumb";
+
+export function updateBreadcrumbsWithItem(
+  breadcrumbs: SdkBreadcrumbItem[],
+  nextItem: SdkBreadcrumbItem,
+): SdkBreadcrumbItem[] {
+  if (breadcrumbs.length === 0) {
+    return [nextItem];
+  }
+
+  const lastItem = breadcrumbs[breadcrumbs.length - 1];
+
+  // When navigating to a dashboard card, do not show in breadcrumbs
+  // for consistency with ad-hoc questions.
+  if (lastItem.type === "dashboard" && nextItem.type === "question") {
+    return breadcrumbs;
+  }
+
+  if (nextItem.type === "collection") {
+    return updateBreadcrumbsWithCollection(breadcrumbs, nextItem);
+  }
+
+  // If a dashboard or question already exist as the last item, replace it.
+  if (lastItem.type === nextItem.type) {
+    return [...breadcrumbs.slice(0, -1), nextItem];
+  }
+
+  return [...breadcrumbs, nextItem];
+}
+
+function updateBreadcrumbsWithCollection(
+  breadcrumbs: SdkBreadcrumbItem[],
+  nextItem: SdkBreadcrumbItem,
+): SdkBreadcrumbItem[] {
+  const existingIndex = breadcrumbs.findIndex(
+    (item) => item.id === nextItem.id && item.type === nextItem.type,
+  );
+
+  // If the collection already exist on the breadcrumb,
+  // remove all breadcrumbs after it.
+  if (existingIndex !== -1) {
+    return breadcrumbs.slice(0, existingIndex + 1);
+  }
+
+  return [...breadcrumbs, nextItem];
```

Comment:
Almost the same logic is defined in `navigateTo` in `enterprise/frontend/src/embedding-sdk/components/private/Breadcrumb/SdkBreadcrumbsProvider.tsx`:

```
 const existingIndex = breadcrumbs.findIndex(
        (b) => b.id === breadcrumb.id && b.type === breadcrumb.type,
      );

      if (existingIndex !== -1) {
        const breadcrumb = breadcrumbs[existingIndex];
        breadcrumb?.onNavigate?.();

        // Remove all breadcrumbs after this item.
        setBreadcrumbs(breadcrumbs.slice(0, existingIndex + 1));
      }
```

The only difference these lines:
```
  const breadcrumb = breadcrumbs[existingIndex];
  breadcrumb?.onNavigate?.();
```

Can we share this code?